src/Streamly/Coreutils/Cp.hs

```diff
@@ -0,0 +1,258 @@
+    , overwriteExisting :: Bool
+    -- ^ If @True@, an existing destination file will be overwritten
+    -- due to copying
+    }
+
+
+-- | Default options for @cp@ are @verbose@ and @overwriteExisting@ to
+-- @True@ and @force@ set to @False@
+--
+-- @since 0.1.0.0
+{-# INLINE defaultCpOptions #-}
+defaultCpOptions :: CpOptions
+defaultCpOptions = CpOptions True False True
+
+
+-------------------------------------------------------------------------------
+-- cp and helper functions for options
+-------------------------------------------------------------------------------
+
+
+-- | Prints paths of source, destination files when @verbose@ is @True@.
+-- Returns @True@ if copying operation was successful.
+-- @False@ is returned in cases where a file by the same name
+-- exists in the destination path and @overwriteExisting@ is set to
+-- @False@ or if the source file does not exist.
+--
+-- @since 0.1.0.0
+{-# INLINE copy #-}
+copy
+    :: CpOptions
+    -> Path Abs File
+    -> Path Abs File
+    -> IO Bool
+copy opt src dst = do
+    let srcFp = fromAbsFile src
+    let dstFp = fromAbsFile dst
+    existDst <- doesFileExist dstFp
+    existSrc <- doesFileExist srcFp
+    if not existSrc
+    then do
+        putStrLn $ "Source file " ++ srcFp ++ " does not exist"
+        return False
+    else if existDst && not (overwriteExisting opt)
+    then do
+        putStrLn $ "overwriteExisting : False " ++ dstFp ++ " already exists"
+        return False
+    else if verbose opt == True
+    then do
+        File.fromChunks dstFp $ File.toChunksWithBufferOf (256*1024) srcFp
+        putStrLn $ srcFp ++ " -> " ++ dstFp
+        return True
+    else do
+        File.fromChunks dstFp $ File.toChunksWithBufferOf (256*1024) srcFp
+        return True
+
+
+-- | Lists the contents (both files and directories) of the directory
+-- as absolute paths.
+-- @Left path@ represents that @path@ is a directory while @Right path@
+-- represents that it is a file.
+--
+-- @since 0.1.0.0
+traverseDir
+    :: IsStream t
+    => Path Abs Dir
+    -> t IO (Either (Path Abs Dir) (Path Abs File))
+traverseDir baseDir = do
+    IP.concatMapTreeWith ahead listContents $ S.yield $ Left baseDir
+
+    where
+
+    listContents :: IsStream t => Path Abs Dir -> t IO (Either (Path Abs Dir) (Path Abs File))
+    listContents dir = do
+        S.fromListM $ map (identify dir) $ contents $ fromAbsDir dir
+
+    identify :: Path Abs Dir -> FilePath -> IO (Either (Path Abs Dir) (Path Abs File))
+    identify dir path = do
+       asFile <- parseRelFile path
+       let filePath = dir </> asFile
+       isFile <- doesFileExist $ fromAbsFile filePath
+       if isFile == True
+       then do
+           return $ Right filePath
+       else do
+           asDir <- parseRelDir path
+           return $ Left $ dir </> asDir
+
+    contents :: FilePath -> [FilePath]
+    contents dir = unsafePerformIO $ listDirectory dir
```

Comment:
No need to have an `unsafePerformIO`, you can change `listContents` appropriately to use `contents :: FilePath -> IO [FilePath]`